### ISSUE TITLE
RFC: Introduction of the minimum required OpenTofu version

### ISSUE THREAD

**Issue by kislerdm** [opentofu/opentofu#962]
### Summary

This RFC introduces the minimal required OpenTofu version and proposes the way to handle mismatch between the OpenTofu 
version used to _write_ the state file and the OpenTofu version used to _update_ it afterward.

### Problem Statement

OpenTofu v1.x [guarantees](https://opentofu.org/docs/language/v1-compatibility-promises) backwards compatibility for a subset of critical language features throughout the 1.x releases. However, _non-forward compatible_ changes could be implemented based on the [community interest](https://github.com/opentofu/opentofu/issues?q=is%3Aopen+is%3Aissue+label%3Af-function) in extending the OpenTofu language. These changes may lead to disturbance of collaborative work relying on the same state to update common infrastructure.

### User-facing description

Let's assume that the resources were provisioned based on the OpenTofu configuration which uses the language features 
introduced in v1.6.1. If OpenTofu v1.6.0 will be subsequently employed to update the infrastructure, the `tofu`  operations will fail. This RFC proposes the functionality to address this problem by updating the `tofu` version during the `init` operation.

### Technical Description

This proposal is influenced by the Go toolchain update introduced in Go v1.21. 

New OpenTofu functionality can be described as following:  

1. OpenTofu identifies the _minimum required version_ during the `init` operation:
    - Read the `required_version` keyword from the configuration module;
    - It not found, read the `terraform_version` value from the state file. 
2. If the found version is greater than the current CLI version, the process will return the _warning_ and prompt the user to update `tofu` version by answering "yes/no":
   - If the user types "no", then
       - OpenTofu will return the _error_ message followed by update instructions, and terminate with the status code 1.
   - If user types "yes", then
     - OpenTofu will make an attempt to install new tofu version:
       1. It will check `PATH` for `tofu161`;
       2. If not found, OpenTofu will attempt to download `tofu v1.6.1` from GitHub Releases to `${HOME}/.tofu/tofu161`.
     - If failed, OpenTofu will return _error_ followed by update instructions and the link to the OpenTofu GitHub issues, and terminate with the status code 1.
     - If succeeded, OpenTofu will alias `tofu161` as `tofu` for this project, and execute the `tofu init` by calling subprocess.

OpenTofu can be configured for implicit upgrade by the command line option `-auto-upgrade`. In this case,  the status code `2` will be returned for the sake of visibility during CI pipelines execution. This status code is selected to match the code returned when running `tofu plan -detailed-exitcode` with non-empty diff.

The diagram compares the OpenTofu behaviour when running the `init` and `plan` operations: before the proposed change -
_Status Quo_, after the change - _Challenger_.

[![](https://mermaid.ink/img/pako:eNqVVF1r2zAU_StCUGghtSU7_nzYYO0GY4wy2r4sHq1iy7aoIxlLapaF_PdJctKkKdmHn66vzrlX9-hIa1iKisIc1p1Yli0ZFLj7UHCp581A-hZctaTrKG_oUHBgvooNtFRMcAezGYnPzwsICnhxMSYuL9-BJY7xA55dCV6zRg8UaMl4A5SoNXjGXuzhH8fgB9L33Wr26DAufgRL_wRlwDEyFMaZ2jJseERAW8LZGbi7ub7Jge4roqhBgDkFBFS0ZNKOMu9E-bQvrmf-vaQD6Aex6FUO7h3tvb8tpy2IC4usCeuQm_-TifYSOMiKSothXCqj4XpdwM9jeDhTATebkbPFgV1dq9pLj-Btj0O81GVJpdxJg_8qzU7LA7Aj92NiS-w7wt8QwV4o8WS3dfNl3BTl1YFxbhVRWoJvWpw0DnoxDjgyA5oV8A_eKeCxe9D_uwf9o3t2shjGaVnQoSz2xF6fl5MGTuCCDgvCKnPd1hZeQNXSBS1gbsKK1kR3qoAF3xgo0UrcrngJczVoOoGjd68ZMfIuXic_VkyJAeY16aRJdoJU1PyuoVr19mo3TCpTsXSC2rweOpNulepl7vt22WuYavXcK8XCl6yy70D7nMV-HMQpCUIaJyGJwrAq5zhL62CK6ypBOCDGvRNIXf-v4zvinpMJ7Am3nX7CPE48FIYBSrMkyqI0nU7gCuZh7E1xFgcBTqdRlCVTU-eXEGYy5CUBjlAYRVGSTOM0TVy1727Rjr35DVveby4?type=png)](https://mermaid.live/edit#pako:eNqVVF1r2zAU_StCUGghtSU7_nzYYO0GY4wy2r4sHq1iy7aoIxlLapaF_PdJctKkKdmHn66vzrlX9-hIa1iKisIc1p1Yli0ZFLj7UHCp581A-hZctaTrKG_oUHBgvooNtFRMcAezGYnPzwsICnhxMSYuL9-BJY7xA55dCV6zRg8UaMl4A5SoNXjGXuzhH8fgB9L33Wr26DAufgRL_wRlwDEyFMaZ2jJseERAW8LZGbi7ub7Jge4roqhBgDkFBFS0ZNKOMu9E-bQvrmf-vaQD6Aex6FUO7h3tvb8tpy2IC4usCeuQm_-TifYSOMiKSothXCqj4XpdwM9jeDhTATebkbPFgV1dq9pLj-Btj0O81GVJpdxJg_8qzU7LA7Aj92NiS-w7wt8QwV4o8WS3dfNl3BTl1YFxbhVRWoJvWpw0DnoxDjgyA5oV8A_eKeCxe9D_uwf9o3t2shjGaVnQoSz2xF6fl5MGTuCCDgvCKnPd1hZeQNXSBS1gbsKK1kR3qoAF3xgo0UrcrngJczVoOoGjd68ZMfIuXic_VkyJAeY16aRJdoJU1PyuoVr19mo3TCpTsXSC2rweOpNulepl7vt22WuYavXcK8XCl6yy70D7nMV-HMQpCUIaJyGJwrAq5zhL62CK6ypBOCDGvRNIXf-v4zvinpMJ7Am3nX7CPE48FIYBSrMkyqI0nU7gCuZh7E1xFgcBTqdRlCVTU-eXEGYy5CUBjlAYRVGSTOM0TVy1727Rjr35DVveby4)

### Alternative: Return warning message with upgrade instructions only

_Pro_:
- Explicit version management.

_Cons_:
- Need to rely on external versions manager. 

### Downsides

The teams would rely on implicit `tofu` update without explicit version maintenance at the level of CI pipeline.

### Unresolved Questions

- Do you find the proposed functionality valuable as addition to existing version managers, e.g. `tfswitch`?
- What risk factors the proposed functionality may cause for your teams?

**Comment by janosdebugs** [comment]
Hey @kislerdm thank you very much for this issue, we have discussed it in the core team. In the past few weeks several tools have popped up that achieve this, see the [Awesome OpenTofu page](https://awesome-opentofu.com/). Additionally, not everyone installs their OpenTofu the same way and simply downloading a binary and then running it may not work due to execution policies. This being a solved problem with external tools and given the complexity of implementing it ourselves, we decided not to take on this issue as part of the OpenTofu core.
